core/src/main/java/jenkins/slaves/DefaultJnlpSlaveReceiver.java

```diff
@@ -47,10 +48,21 @@ public boolean owns(String clientName) {
     public void afterProperties(@NonNull JnlpConnectionState event) {
         String clientName = event.getProperty(JnlpConnectionState.CLIENT_NAME_KEY);
         SlaveComputer computer = (SlaveComputer) Jenkins.getInstance().getComputer(clientName);
-        if (computer == null || !(computer.getLauncher() instanceof JNLPLauncher)) {
-            event.reject(new ConnectionRefusalException(String.format("%s is not a JNLP agent", clientName)));
+        if (computer == null) {
+            event.reject(new ConnectionRefusalException(String.format("%s is not a registered JNLP agent", clientName)));
             return;
         }
+        
+        final ComputerLauncher launcher = computer.getLauncher();
+        if (!(launcher instanceof JNLPLauncher) && LOGGER.isLoggable(Level.FINE)) {
```

Comment:
If you are going to do this rather than walk the delegation chain, you need to reject a connection over JNLP where a connection is already established and connected.